File: lib/changelog/src/rewrite.ts
import type { EntityImpact, PatchStats } from '@deadlog/utils';
import {
	applyUpdate,
	writeEnrichmentBlock,
	type EnrichmentUpdate,
	type EntityEnrichment
} from './entityEnrichment';
import { changeKey, parseStructure } from './extract';
import { writeStatsNode } from './impactBlock';
import type { EntityBlock } from './schema';

function spliceStats(
	lines: string[],
	metadataLines: [number, number] | null,
	stats: PatchStats
): void {
	if (!metadataLines) throw new Error('No root attr block to hold the stats node');
	const [start, end] = metadataLines;
	const first = lines.findIndex(
		(line, index) => index > start && index < end && /^stats\b/.test(line)
	);
	if (first === -1) {
		lines.splice(end, 0, ...writeStatsNode(stats));
		return;
	}
	const last = lines.findIndex((line, index) => index > first && line === '}');
	if (last === -1 || last >= end) throw new Error('Unterminated stats node');
	lines.splice(first, last - first + 1, ...writeStatsNode(stats));
}

export async function spliceEntityBlocks(
	source: string,
	pick: (block: EntityBlock) => EnrichmentUpdate | undefined,
	stats?: PatchStats
): Promise<string> {
	if (source.includes('\r')) throw new Error('CRLF line endings are not supported');

	const lines = source.split('\n');
	const { blocks, metadataLines } = await parseStructure(source);

	for (const block of [...blocks].sort((a, b) => b.fenceLine - a.fenceLine)) {
		const update = pick(block);
		if (!update) continue;
		const [start, end] = block.attributeLines ?? [block.fenceLine + 1, block.fenceLine];
		lines.splice(
			start,
			end - start + 1,
			...writeEnrichmentBlock(applyUpdate(block.enrichment, update))
		);
	}
	if (stats) spliceStats(lines, metadataLines, stats);
	return lines.join('\n');
}

export function spliceImpactBlocks(
	source: string,
	pick: (block: EntityBlock) => EntityImpact | null | undefined,
	stats?: PatchStats
): Promise<string> {
	return spliceEntityBlocks(
		source,
		(block) => {
			const impact = pick(block);
			return impact === undefined ? undefined : { impact };
		},
		stats
	);
}

export async function carryEnrichment(previous: string, next: string): Promise<string> {
	const recorded = new Map<string, EntityEnrichment>();
	const before = await parseStructure(previous);
	for (const block of before.blocks) {
		if (Object.keys(block.enrichment).length > 0) {
			recorded.set(changeKey(block.type, block.name), block.enrichment);
		}
	}
	if (recorded.size === 0) return next;

	const after = await parseStructure(next);
	return spliceEntityBlocks(
		next,
		(block) =>
			block.attributeLines ? undefined : recorded.get(changeKey(block.type, block.name)),
		after.stats ? undefined : (before.stats ?? undefined)
	);
}

File: crates/changelog/src/rewrite.rs
use std::collections::HashMap;

use anyhow::{Result, bail};
use deadlog_model::{EntityImpact, PatchStats};

use crate::enrichment::{EnrichmentUpdate, EntityEnrichment, apply_update, write_enrichment_block};
use crate::extract::{change_key, parse_structure};
use crate::impact_block::write_stats_node;
use crate::schema::EntityBlock;

fn is_stats_line(line: &str) -> bool {
    line.strip_prefix("stats").is_some_and(|rest| !rest.starts_with(|c: char| c.is_ascii_alphanumeric() || c == '_'))
}

fn splice_stats(lines: &mut Vec<String>, metadata_lines: Option<(usize, usize)>, stats: &PatchStats) -> Result<()> {
    let Some((start, end)) = metadata_lines else {
        bail!("No root attr block to hold the stats node");
    };
    let first = lines.iter().enumerate().position(|(index, line)| index > start && index < end && is_stats_line(line));
    let Some(first) = first else {
        lines.splice(end..end, write_stats_node(stats));
        return Ok(());
    };
    let last = lines.iter().enumerate().position(|(index, line)| index > first && line == "}");
    match last {
        Some(last) if last < end => {
            lines.splice(first..=last, write_stats_node(stats));
            Ok(())
        }
        _ => bail!("Unterminated stats node"),
    }
}

/// Rewrites the attr block under each entity fence `pick` answers for, and the root
/// stats node when given one. Everything else keeps its bytes.
pub fn splice_entity_blocks(
    source: &str,
    mut pick: impl FnMut(&EntityBlock) -> Option<EnrichmentUpdate>,
    stats: Option<&PatchStats>,
) -> Result<String> {
    if source.contains('\r') {
        bail!("CRLF line endings are not supported");
    }

    let mut lines: Vec<String> = source.split('\n').map(str::to_string).collect();
    let parsed = parse_structure(source)?;
    let mut blocks: Vec<&EntityBlock> = parsed.blocks.iter().collect();
    blocks.sort_by(|a, b| b.fence_line.cmp(&a.fence_line));

    for block in blocks {
        let Some(update) = pick(block) else { continue };
        let (start, end) = block.attribute_lines.unwrap_or((block.fence_line + 1, block.fence_line));
        let written = write_enrichment_block(&apply_update(&block.enrichment, &update));
        lines.splice(start..end + 1, written);
    }
    if let Some(stats) = stats {
        splice_stats(&mut lines, parsed.metadata_lines, stats)?;
    }
    Ok(lines.join("\n"))
}

/// `pick` answers `None` to leave a block alone and `Some(None)` to clear its impact.
pub fn splice_impact_blocks(
    source: &str,
    mut pick: impl FnMut(&EntityBlock) -> Option<Option<EntityImpact>>,
    stats: Option<&PatchStats>,
) -> Result<String> {
    splice_entity_blocks(
        source,
        |block| pick(block).map(|impact| EnrichmentUpdate { impact: Some(impact), ..Default::default() }),
        stats,
    )
}

/// Restores what `previous` recorded — entity blocks and the stats node — onto a
/// regeneration of the same patch, matching entities by heading name.
pub fn carry_enrichment(previous: &str, next: &str) -> Result<String> {
    let before = parse_structure(previous)?;
    let recorded: HashMap<String, &EntityEnrichment> = before
        .blocks
        .iter()
        .filter(|block| !block.enrichment.is_empty())
        .map(|block| (change_key(block.kind.as_str(), &block.name), &block.enrichment))
        .collect();
    if recorded.is_empty() {
        return Ok(next.to_string());
    }

    let after = parse_structure(next)?;
    let stats = if after.stats.is_some() { None } else { before.stats.as_ref() };
    splice_entity_blocks(
        next,
        |block| {
            if block.attribute_lines.is_some() {
                return None;
            }
            recorded.get(&change_key(block.kind.as_str(), &block.name)).map(|enrichment| (*enrichment).clone().into())
        },
        stats,
    )
}
